README.md
SolidPython
-----------
- [SolidPython:  OpenSCAD for Python](#solidpython--openscad-for-python)
- [Advantages](#advantages)
- [Installing SolidPython](#installing-solidpython)
- [Using SolidPython](#using-solidpython)
- [Example Code](#example-code)
- [Extra syntactic sugar](#extra-syntactic-sugar)
	- [Basic operators](#basic-operators)
    - [First-class Negative Space (Holes)](#first-class-negative-space-holes)
    - [Animation](#animation)
- [solid.utils](#solidutils)
	- [Directions: (up, down, left, right, forward, back) for arranging things:](#directions-up-down-left-right-forward-back-for-arranging-things)
	- [Arcs](#arcs)
    - [Offsets](#offsets)
    - [Extrude Along Path](#extrude_along_path)
	- [Basic color library](#basic-color-library)
	- [Bill Of Materials](#bill-of-materials)
- [solid.screw_thread](#solidscrew_thread)
- [Contact](#contact)
- [License](#license)

**Table of Contents**  *generated with [DocToc](http://doctoc.herokuapp.com/)*


# SolidPython:  OpenSCAD for Python<a id="solidpython--openscad-for-python"></a>

SolidPython is a generalization of Phillip Tiefenbacher's openscad module, 
found on [Thingiverse](http://www.thingiverse.com/thing:1481).  It generates valid OpenSCAD 
code from Python code with minimal overhead.  Here's a simple example:
    
This Python code: 

    from solid import *
    d = difference()(
        cube(10),
        sphere(15)
    )
    print(scad_render(d))
 

Generates this OpenSCAD code:

    difference(){
        cube(10);
        sphere(15);
    }

That doesn't seem like such a savings, but the following SolidPython code is a 
lot shorter (and I think a lot clearer) than the SCAD code it compiles to:

    d = cube(5) + right(5)(sphere(5)) - cylinder(r=2, h=6)

Generates this OpenSCAD code:

    difference(){
        union(){
            cube(5);
            translate( [5, 0,0]){
                sphere(5);
            }
        }
        cylinder(r=2, h=6);
    }

# Advantages<a id="advantages"></a>
Because you're using Python, a lot of things are easy that would be hard or 
impossible in pure OpenSCAD.  Among these are:

* built-in dictionary types
* mutable, slice-able list and string types
* recursion
* external libraries (images! 3D geometry!  web-scraping! ...)

# Installing SolidPython<a id="installing-solidpython"></a>
*   Install via [PyPI](python setup.py sdist bdist_wininst upload):

        pip install solidpython
        
    (You may need to use `sudo pip install solidpython`, depending on your environment.)
    
*   **OR:** Download SolidPython (Click [here](https://github.com/SolidCode/SolidPython/archive/master.zip) to download directly, or use git to pull it all down)

    (Note that SolidPython also depends on the [PyEuclid](http://pypi.python.org/pypi/euclid) Vector math library, installable via `sudo pip install euclid`)
    
    *   Unzip the file, probably in ~/Downloads/SolidPython-master
    *   In a terminal, cd to location of file:
  
            cd ~/Downloads/SolidPython-master

    *   Run the install script: 

            sudo python setup.py --install

# Using SolidPython<a id="using-solidpython"></a>
*   Include SolidPython at the top of your Python file:

        from solid import *
        from solid.utils import *  # Not required, but the utils module is useful
*   To include other scad code, call ```use("/path/to/scadfile.scad")``` or ```include("/path/to/scadfile.scad")```. This is identical to what you would
    do in OpenSCAD. 
*   OpenSCAD uses curly-brace blocks ({}) to create its tree.  SolidPython uses
    parentheses with comma-delimited lists.
    __OpenSCAD:__
    
        difference(){
            cube(10);
            sphere(15);
        }

    __SolidPython:__
    
        d = difference()(
            cube(10),  # Note the comma between each element!
            sphere(15)
        )
           
*   Call ```scad_render(py_scad_obj)``` to generate SCAD code. This returns a string of valid OpenSCAD code.
*   *or*: call ```scad_render_to_file(py_scad_obj, filepath)``` to
    store that code in a file.  
*   If 'filepath' is open in the OpenSCAD IDE and Design =>
    'Automatic Reload and Compile' is checked (in the OpenSCAD IDE), calling
    ```scad_render_to_file()``` from Python will load the object in
    the IDE.
*   Alternately, you could call OpenSCAD's command line and render straight 
    to STL.   
    
# Example Code<a id="example-code"></a>
The best way to learn how SolidPython works is to look at the included example code. 
If you've installed SolidPython, the  following line of Python will print(the location of )
the examples directory:

        import os, solid; print(os.path.dirname(solid.__file__) + '/examples')
        
Or browse the example code on Github [here](https://github.com/SolidCode/SolidPython/tree/master/solid/examples)

Adding your own code to the example file [`solid/examples/solidpython_template.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/examples/solidpython_template.py) will make some of the setup easier.

# Extra syntactic sugar<a id="extra-syntactic-sugar"></a>
### Basic operators<a id="basic-operators"></a>
Following Elmo Mäntynen's suggestion, SCAD objects override 
the basic operators + (union), - (difference), and * (intersection).
So

    c = cylinder(r=10, h=5) + cylinder(r=2, h=30)
is the same as:

    c = union()(
        cylinder(r=10, h=5),
        cylinder(r=2, h=30)
    )

Likewise:

    c = cylinder(r=10, h=5)
    c -= cylinder(r=2, h=30)

is the same as:

    c = difference()(
        cylinder(r=10, h=5),
        cylinder(r=2, h=30)
    )

### First-class Negative Space (Holes)<a id="first-class-negative-space-holes"></a>
OpenSCAD requires you to be very careful with the order in which you add or 
subtract objects. SolidPython's `hole()` function makes this process easier.

Consider making a joint where two pipes come together.  In OpenSCAD you need to
make two cylinders, union them, then make two smaller cylinders, union
them, then subtract the smaller from the larger. 

Using hole(), you can make a pipe, specify that its center should remain open,
and then add two pipes together knowing that the central void area will stay 
empty no matter what other objects are added to that structure.

Example: 

    outer = cylinder(r=pipe_od, h=seg_length)
    inner = cylinder(r=pipe_id, h=seg_length)
    pipe_a = outer - hole()(inner)

Once you've made something a hole, eventually you'll want to put something, 
like a bolt, into it.  To do this, we need to specify that there's a given 
'part' with a hole and that other parts may occupy the space in that hole.
This is done with the `part()` function.  

See [`solid/examples/hole_example.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/examples/hole_example.py) for the complete picture.

### Animation<a id="animation"></a>
OpenSCAD has a special variable, `$t`, that can be used to animate motion.
SolidPython can do this, too, using the special function `scad_render_animated_file()`.  

See [`solid/examples/animation_example.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/examples/animation_example.py) for more details.

# solid.utils<a id="solidutils"></a>
SolidPython includes a number of useful functions in [`solid/utils.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/utils.py).  Currently these include:
    
### Directions: (up, down, left, right, forward, back) for arranging things:<a id="directions-up-down-left-right-forward-back-for-arranging-things"></a>
    
    up(10)(
        cylinder()
    )

seems a lot clearer to me than:

    translate( [0,0,10])(
        cylinder()
    )
    

I took this from someone's SCAD work and have lost track of the original author.  
My apologies.
    
### Arcs<a id="arcs"></a>
I've found this useful for fillets and rounds.

    arc(rad=10, start_degrees=90, end_degrees=210)

draws an arc of radius 10 counterclockwise from 90 to 210 degrees. 

    arc_inverted(rad=10, start_degrees=0, end_degrees=90) 

draws the portion of a 10x10 square NOT in a 90 degree circle of radius 10.
This is the shape you need to add to make fillets or remove to make rounds.

### Offsets<a id="offsets"></a>
To offset a set of points in one direction or another (inside or outside a closed 
figure, for example) use `solid.utils.offset_points(point_arr, offset, inside=True)`

Note that, for a non-convex figure, inside and outside may be non-intuitive.  The 
simple solution is to manually check that your offset is going in the direction you
intend, and change the boolean value of `inside` if you're not happy. 

See the code for futher explanation. Improvements on the inside/outside algorithm would be welcome.

### Extrude Along Path<a id="extrude_along_path"></a>
`solid.utils.extrude_along_path(shape_pts, path_pts, scale_factors=None)`

See [`solid/examples/path_extrude_example.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/examples/path_extrude_example.py) for use.

    
### Basic color library<a id="basic-color-library"></a>
You can change an object's color by using the OpenSCAD ```color([rgba_array])``` function:

    transparent_blue = color([0,0,1, 0.5])(cube(10))  # Specify with RGB[A]
    red_obj = color(Red)(cube(10))                    # Or use predefined colors

These colors are pre-defined in solid.utils:

<table border="0" cellspacing="5" cellpadding="0">
    <tr><td>Red        </td><td>Green      </td><td>Blue       </td></tr>
    <tr><td>Cyan       </td><td>Magenta    </td><td>Yellow     </td></tr>
    <tr><td>Black      </td><td>White      </td><td>Transparent</td></tr>
    <tr><td>Oak        </td><td>Pine       </td><td>Birch      </td></tr>
    <tr><td>Iron       </td><td>Steel      </td><td>Stainless  </td></tr>
    <tr><td>Aluminum   </td><td>Brass      </td><td>BlackPaint </td></tr>
    <tr><td>FiberBoard </td></tr>
</table>

They're a conversion of the materials in the [MCAD OpenSCAD library](https://github.com/openscad/MCAD), as seen [here] (https://github.com/openscad/MCAD/blob/master/materials.scad).
    
### Bill Of Materials<a id="bill-of-materials"></a>
Put ```@bom_part()``` before any method that defines a part, then 
call ```bill_of_materials()``` after the program is run, and all parts will be 
counted, priced and reported. 

The example file [`solid/examples/bom_scad.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/examples/bom_scad.py) illustrates this. Check it out.

## solid.screw_thread<a id="solidscrew_thread"></a>
solid.screw_thread includes a method, thread() that makes internal and external 
screw threads.  

See [`solid/examples/screw_thread_example.py`](https://github.com/SolidCode/SolidPython/blob/master/solid/examples/screw_thread_example.py) for more details.

# Contact<a id="contact"></a>
Enjoy, and please send any questions or bug reports to me at ```evan_t_jones@mac.com```. 

Cheers!

Evan

# License<a id="license"></a>
This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

[Full text of the license](http://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt).


CHANGES.rst
0.0.2 (unreleased)
==================

- Nothing changed yet.


0.0.1 (2016-08-17)
==================

- forked github.com/SolidCode/SolidPython
- introduced VERSION

pyopenscad.py
#! /usr/bin/python
# -*- coding: utf-8 -*-

#    Simple Python OpenSCAD Code Generator
#    Copyright (C) 2009    Philipp Tiefenbacher <wizards23@gmail.com>
#    Amendments & additions, (C) 2011 Evan Jones <evan_t_jones@mac.com>
#
#   License: LGPL 2.1 or later
#


import os, sys, re
import inspect

openscad_builtins = [
    # 2D primitives
    {'name': 'polygon',         'args': ['points', 'paths'], 'kwargs': []} ,
    {'name': 'circle',          'args': [],         'kwargs': ['r', 'segments']} ,
    {'name': 'square',          'args': [],         'kwargs': ['size', 'center']} ,
    
    # 3D primitives
    {'name': 'sphere',          'args': [],         'kwargs': ['r', 'segments']} ,
    {'name': 'cube',            'args': [],         'kwargs': ['size', 'center']} ,
    {'name': 'cylinder',        'args': [],         'kwargs': ['r','h','r1', 'r2', 'center', 'segments']}  ,
    {'name': 'polyhedron',      'args': ['points', 'triangles' ], 'kwargs': ['convexity']} ,
    
    # Boolean operations
    {'name': 'union',           'args': [],         'kwargs': []} ,
    {'name': 'intersection',    'args': [],         'kwargs': []} ,
    {'name': 'difference',      'args': [],         'kwargs': []} ,
    
    # Transforms
    {'name': 'translate',       'args': [],         'kwargs': ['v']} ,
    {'name': 'scale',           'args': [],         'kwargs': ['v']} ,
    {'name': 'rotate',          'args': [],         'kwargs': ['a', 'v']} ,
    {'name': 'mirror',          'args': ['v'],      'kwargs': []},
    {'name': 'multmatrix',      'args': ['n'],      'kwargs': []},
    {'name': 'color',           'args': ['c'],      'kwargs': []},
    {'name': 'minkowski',       'args': [],         'kwargs': []},
    {'name': 'hull',            'args': [],         'kwargs': []},
    {'name': 'render',          'args': [],         'kwargs': ['convexity']}, 
        
    # 2D to 3D transitions
    {'name': 'linear_extrude',  'args': [],         'kwargs': ['height', 'center', 'convexity', 'twist','slices']} ,
    {'name': 'rotate_extrude',  'args': [],         'kwargs': ['convexity']} ,
    {'name': 'dxf_linear_extrude', 'args': ['file'], 'kwargs': ['layer', 'height', 'center', 'convexity', 'twist', 'slices']} ,
    {'name': 'projection',      'args': [],         'kwargs': ['cut']} ,
    
    # Import/export
    {'name': 'import_stl',      'args': ['filename'], 'kwargs': ['convexity']} ,
    
    # Modifiers; These are implemented by calling e.g. 
    #   obj.set_modifier( '*') or 
    #   obj.set_modifier('disable') 
    # on  an existing object.
    # {'name': 'background',      'args': [],         'kwargs': []},     #   %{}
    # {'name': 'debug',           'args': [],         'kwargs': []} ,    #   #{}
    # {'name': 'root',            'args': [],         'kwargs': []} ,    #   !{}
    # {'name': 'disable',         'args': [],         'kwargs': []} ,    #   *{}
    
    {'name': 'intersection_for', 'args': ['n'],     'kwargs': []}  ,    #   e.g.: intersection_for( n=[1..6]){}
    
    # Unneeded
    {'name': 'assign',          'args': [],         'kwargs': []}   # Not really needed for Python.  Also needs a **args argument so it accepts anything
]

# Some functions need custom code in them; put that code here
builtin_literals = {
    'polygon': '''class polygon( openscad_object):
        def __init__( self, points, paths=None):
            if not paths:
                paths = [ range( len( points))]
            openscad_object.__init__( self, 'polygon', {'points':points, 'paths': paths})
        
'''

}

# ===============
# = Including OpenSCAD code =
# ===============
def use( scad_file_path, use_not_include=True):
    '''
    TODO:  doctest needed
    '''
    # Opens scad_file_path, parses it for all usable calls, 
    # and adds them to caller's namespace
    
    # TODO: add something along the lines of PYTHONPATH for scad files?
    # That way you could 'use( a_file.scad)' without using an absolute
    # path or having the library in the same directory
    try:
        module = open( scad_file_path)
        contents = module.read()
        module.close()
    except Exception, e:
        raise Exception( "Failed to import SCAD module '%(scad_file_path)s' with error: %(e)s "%vars())
    
    # Once we have a list of all callables and arguments, dynamically
    # add openscad_object subclasses for all callables to the calling module's
    # namespace.
    symbols_dicts = extract_callable_signatures( scad_file_path)
    
    for sd in symbols_dicts:
        class_str = new_openscad_class_str( sd['name'], sd['args'], sd['kwargs'], scad_file_path, use_not_include)
        exec class_str in calling_module().__dict__
    
    return True

def include( scad_file_path):
    return use( scad_file_path, use_not_include=False)


# =========================================
# = Rendering Python code to OpenSCAD code=
# =========================================
def scad_render( scad_object, file_header=''):
    # Find the root of the tree, calling x.parent until there is none
    root = scad_object
    
    # Scan the tree for all instances of 
    # included_openscad_object, storing their strings
    def find_include_strings( obj):
        include_strings = set()
        if isinstance( obj, included_openscad_object):
            include_strings.add( obj.include_string )
        for child in obj.children:
            include_strings.update( find_include_strings( child))
        return include_strings
    
    include_strings = find_include_strings( root)
    
    # and render the string
    includes = ''.join(include_strings) + "\n"
    scad_body = root._render()
    return file_header + includes + scad_body

def scad_render_to_file( scad_object, filepath=None, file_header='', include_orig_code=False):
    rendered_string = scad_render( scad_object, file_header)
    
    calling_file = os.path.abspath( calling_module().__file__) 
    
    if include_orig_code:
        # Once a SCAD file has been created, it's difficult to reconstruct
        # how it got there, since it has no variables, modules, etc.  So, include
        # the Python code that generated the scad code as comments at the end of 
        # the SCAD code
        pyopenscad_str = open(calling_file, 'r').read()
    
        pyopenscad_str = '''
/***********************************************
******      PyOpenSCAD code:       *************
************************************************
 
%(pyopenscad_str)s 
 
***********************************************/
                            
'''%vars()        
        rendered_string += pyopenscad_str
    
    # This write is destructive, and ought to do some checks that the write
    # was successful.
    # If filepath isn't supplied, place a .scad file with the same name
    # as the calling module next to it
    if not filepath:
        filepath = os.path.splitext( calling_file)[0] + '.scad'
    
    f = open( filepath,"w")
    f.write( rendered_string)
    f.close


# =========================
# = Internal Utilities    =
# =========================
class openscad_object( object):
    def __init__(self, name, params):
        self.name = name
        self.params = params
        self.children = []
        self.modifier = ""
        self.parent= None
    
    def set_modifier(self, m):
        # Used to add one of the 4 single-character modifiers: #(debug)  !(root) %(background) or *(disable)
        string_vals = { 'disable':      '*',
                        'debug':        '#',
                        'background':   '%',
                        'root':         '!',
                        '*':'*',
                        '#':'#',
                        '%':'%',
                        '!':'!'}
         
        self.modifier = string_vals.get(m.lower(), '')
        return self
    
    def _render(self):
        '''
        NOTE: In general, you won't want to call this method. For most purposes,
        you really want scad_render(), 
        Calling obj._render won't include necessary 'use' or 'include' statements
        '''        
        s = "\n" + self.modifier + self.name + "("
        first = True
            
        # OpenSCAD doesn't have a 'segments' argument, but it does 
        # have '$fn'.  Swap one for the other
        if 'segments' in self.params:
            self.params['$fn'] = self.params.pop('segments')
            
        valid_keys = self.params.keys()
            
        # intkeys are the positional parameters
        intkeys = filter(lambda x: type(x)==int, valid_keys)
        intkeys.sort()
        
        # named parameters
        nonintkeys = filter(lambda x: not type(x)==int, valid_keys)
        
        for k in intkeys+nonintkeys:
            v = self.params[k]
            if v == None:
                continue
            
            if not first:
                s += ", "
            first = False
            
            if type(k)==int:
                s += py2openscad(v)
            else:
                s += k + " = " + py2openscad(v)
                
        s += ")"
        if self.children != None and len(self.children) > 0:
            s += " {"
            for child in self.children:
                s += indent(child._render())
            s += "\n}"
        else:
            s += ";"
        return s
    
    def add(self, child):
        '''
        if child is a single object, assume it's an openscad_object and 
        add it to self.children
        
        if child is a list, assume its members are all openscad_objects and
        add them all to self.children
        '''
        if isinstance( child, list) or isinstance( child, tuple):
            [self.add( c) for c in child]
        else:
            self.children.append(child)
            child.set_parent( self)
        return self
    
    def set_parent( self, parent):
        self.parent = parent
    
    def add_param(self, k, v):
        self.params[k] = v
        return self
    
    def copy( self):
        # Provides a copy of this object and all children, 
        # but doesn't copy self.parent, meaning the new object belongs
        # to a different tree
        other = openscad_object( self.name, self.params)
        for c in self.children:
            other.add( c.copy())
        return other
    
    def __call__( self, *args):
        '''
        Adds all objects in args to self.  This enables OpenSCAD-like syntax,
        e.g.:
        union()(
            cube(),
            sphere()
        )
        '''
        return self.add(args)
    
    def __add__(self, x):
        '''
        This makes u = a+b identical to:
        u = union()( a, b )
        '''
        return union()(self, x)
    
    def __sub__(self, x):
        '''
        This makes u = a - b identical to:
        u = difference()( a, b )
        '''        
        return difference()(self, x)
    
    def __mul__(self, x):
        '''
        This makes u = a * b identical to:
        u = intersection()( a, b )
        '''        
        return intersection()(self, x)
    

class included_openscad_object( openscad_object):
    '''
    Identical to openscad_object, but each subclass of included_openscad_object
    represents imported scad code, so each instance needs to store the path
    to the scad file it's included from.
    '''
    def __init__( self, name, params, include_file_path, use_not_include=False):
        self.include_file_path = include_file_path
        if use_not_include:
            self.include_string = 'use <%s>\n'%self.include_file_path
        else:
            self.include_string = 'include <%s>\n'%self.include_file_path
        
        openscad_object.__init__( self, name, params)
    


def calling_module():
    '''
    Returns the module *2* back in the frame stack.  That means:
    code in module A calls code in module B, which asks calling_module()
    for module A.
    
    Got that?
    '''
    frm = inspect.stack()[2]
    calling_mod = inspect.getmodule( frm[0])
    return calling_mod

def new_openscad_class_str( class_name, args=[], kwargs=[], include_file_path=None, use_not_include=True):
    args_str = ''
    args_pairs = ''
    
    for arg in args:
        args_str += ', '+arg
        args_pairs += "'%(arg)s':%(arg)s, "%vars()
        
    # kwargs have a default value defined in their SCAD versions.  We don't 
    # care what that default value will be (SCAD will take care of that), just
    # that one is defined.
    for kwarg in kwargs:
        args_str += ', %(kwarg)s=None'%vars()
        args_pairs += "'%(kwarg)s':%(kwarg)s, "%vars()
        
    if include_file_path:
        result = '''class %(class_name)s( included_openscad_object):
        def __init__(self%(args_str)s):
            included_openscad_object.__init__(self, '%(class_name)s', {%(args_pairs)s }, include_file_path='%(include_file_path)s', use_not_include=%(use_not_include)s )
        
'''%vars()
    else:
        result = '''class %(class_name)s( openscad_object):
        def __init__(self%(args_str)s):
            openscad_object.__init__(self, '%(class_name)s', {%(args_pairs)s })
        
'''%vars()
    return result

def py2openscad(o):
    if type(o) == bool:
        return str(o).lower()
    if type(o) == float:
        return "%.10f" % o
    if type(o) == list:
        s = "["
        first = True
        for i in o:
            if not first:
                s +=    ", "
            first = False
            s += py2openscad(i)
        s += "]"
        return s
    if type(o) == str:
        return '"' + o + '"'
    return str(o)

def indent(s):
    return s.replace("\n", "\n\t")


# ===========
# = Parsing =
# ===========
def extract_callable_signatures( scad_file_path):
    scad_code_str = open(scad_file_path).read()
    return parse_scad_callables( scad_code_str)

def parse_scad_callables( scad_code_str): 
    """>>> test_str = '''module hex (width=10, height=10,    
...     flats= true, center=false){}
...     function righty (angle=90) = 1;
...     function lefty( avar) = 2;
...     module more( a=[something, other]) {}
...     module pyramid(side=10, height=-1, square=false, centerHorizontal=true, centerVertical=false){}
...     module no_comments( arg=10,   //test comment
...                         other_arg=2, /* some extra comments
...                         on empty lines */
...                         last_arg=4){}
...    module float_arg( arg=1.0){}
... '''
>>> parse_scad_callables( test_str)
[{'args': [], 'name': 'hex', 'kwargs': ['width', 'height', 'flats', 'center']}, {'args': [], 'name': 'righty', 'kwargs': ['angle']}, {'args': ['avar'], 'name': 'lefty', 'kwargs': []}, {'args': [], 'name': 'more', 'kwargs': ['a']}, {'args': [], 'name': 'pyramid', 'kwargs': ['side', 'height', 'square', 'centerHorizontal', 'centerVertical']}, {'args': [], 'name': 'no_comments', 'kwargs': ['arg', 'other_arg', 'last_arg']}, {'args': [], 'name': 'float_arg', 'kwargs': ['arg']}]
>>> """         
    callables = []
    
    # Note that this isn't comprehensive; tuples or nested data structures in 
    # a module definition will defeat it.  
    
    # Current implementation would throw an error if you tried to call a(x, y) 
    # since Python would expect a( x);  OpenSCAD itself ignores extra arguments, 
    # but that's not really preferable behavior 
    
    # TODO:  write a pyparsing grammar for OpenSCAD, or, even better, use the yacc parse grammar
    # used by the language itself.  -ETJ 06 Feb 2011   
           
    no_comments_re = r'(?mxs)(//.*?\n|/\*.*?\*/)'
    
    # Also note: this accepts: 'module x(arg) =' and 'function y(arg) {', both of which are incorrect syntax
    mod_re  = r'(?mxs)^\s*(?:module|function)\s+(?P<callable_name>\w+)\s*\((?P<all_args>.*?)\)\s*(?:{|=)'
    
    # This is brittle.  To get a generally applicable expression for all arguments,
    # we'd need a real parser to handle nested-list default args or parenthesized statements.  
    # For the moment, assume a maximum of one square-bracket-delimited list 
    args_re = r'(?mxs)(?P<arg_name>\w+)(?:\s*=\s*(?P<default_val>[\w.-]+|\[.*\]))?(?:,|$)'
             
    # remove all comments from SCAD code
    scad_code_str = re.sub(no_comments_re,'', scad_code_str)
    # get all SCAD callables
    mod_matches = re.finditer( mod_re, scad_code_str)
    
    for m in mod_matches:
        callable_name = m.group('callable_name')
        args = []
        kwargs = []        
        all_args = m.group('all_args')
        if all_args:
            arg_matches = re.finditer( args_re, all_args)
            for am in arg_matches:
                arg_name = am.group('arg_name')
                if am.group('default_val'):
                    kwargs.append( arg_name)
                else:
                    args.append( arg_name)
        
        callables.append( { 'name':callable_name, 'args': args, 'kwargs':kwargs})
        
    return callables


# Dynamically add all builtins to this namespace on import
for sym_dict in openscad_builtins:
    # entries in 'builtin_literals' override the entries in 'openscad_builtins'
    if sym_dict['name'] in builtin_literals:
        class_str = builtin_literals[ sym_dict['name']]
    else:
        class_str = new_openscad_class_str( sym_dict['name'], sym_dict['args'], sym_dict['kwargs'])
    
    exec class_str 
    
